### PR TITLE
[14.0][ADD] mail_thread_create_nolog: display create msg on the fly

### PR DESCRIPTION
**Pull request by sebalix**
This module will block the creation in the database of 'Record created' messages on data models inheriting from `mail.thread`, but will instead generate this message on the fly when Odoo displays the messages of a record.

This allows to reduce the size of the `mail_message` table.

### CODE REVIEW

File: mail_thread_create_nolog/models/mail_message.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            res.append(create_message_values)
+        return res
+
+    def _get_create_message_id(self, domain):
+        # Return the negative version of the greatest message ID
+        # of the current record
+        return -self.search(domain, limit=1, order="id DESC").id
+
+    @api.model
+    def _messages_for_record(self, domain):
```
Comment (TDu) [inline mail_thread_create_nolog/models/mail_message.py:56]: Could you explain what it does ?

**Review by TDu — COMMENTED** [review]
It needs some unit tests

File: mail_thread_create_nolog/models/mail_message.py (hunk 1 above)
Comment (sebalix) [reply mail_thread_create_nolog/models/mail_message.py:56]: I'll add a docstring yes.
It gets the record from the domain, domain used by Odoo web client to get the messages for that record when it displays the chatter. But the `message_fetch` method is also used to retrieve messages of channels, outside of any record context. In that case, no record = no fake-create-message.

**Comment by sebalix** [comment]
@TDu docstring, comments and tests added

File: mail_thread_create_nolog/models/mail_thread.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+from odoo import api, models
+
+
+class MailThread(models.AbstractModel):
+    _inherit = "mail.thread"
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        # Overloaded to not create 'Record created' messages that increase
```
Comment (simahawk) [inline mail_thread_create_nolog/models/mail_thread.py]:
```suggestion
        # Overridden to not create 'Record created' messages that increase
```

File: mail_thread_create_nolog/models/mail_thread.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+        # the size of 'mail_message' table for little value.
+        # Instead a message will be generated on the fly when the Odoo client
+        # will retrieve the list of messages for a given record.
+        return super(MailThread, self.with_context(mail_create_nolog=True)).create(
```
Comment (simahawk) [inline mail_thread_create_nolog/models/mail_thread.py:16]: I would make this configurable via UI and maybe by model. Worth to add it to the ROADMAP

File: mail_thread_create_nolog/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
+# Copyright 2023 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+{
+    "name": "Mail Thread - Fake create log",
+    "summary": "Display a fake (non-stored) create log in the chatter.",
+    "version": "14.0.1.0.0",
+    "category": "Mail",
+    "website": "https://github.com/OCA/social",
```
Comment (simahawk) [inline mail_thread_create_nolog/__manifest__.py:8]: dev status: alpha ;)

**Review by simahawk — APPROVED** [review]
LG at first implementation.
I would spend some more words on the to explain the hack tho.

File: mail_thread_create_nolog/__manifest__.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# Copyright 2023 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+{
+    "name": "Mail Thread - Fake create log",
```
Comment (mmequignon) [inline mail_thread_create_nolog/__manifest__.py:4]: is it nolog or fake logs?
I would align module / manifest names

File: mail_thread_create_nolog/models/mail_message.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+                    "author_id": (author.id, author.name),
+                }
+            )
+            res.append(create_message_values)
+        return res
+
+    def _get_create_message_id(self, domain):
+        # Return the negative version of the greatest message ID
+        # of the current record
+        return -self.search(domain, limit=1, order="id DESC").id
```
Comment (mmequignon) [inline mail_thread_create_nolog/models/mail_message.py:53]: if search returns an empty recordset, you'll return `-False` which is `0`.
Is it ok?

File: mail_thread_create_nolog/models/mail_message.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+    @api.model
+    def _messages_for_record(self, domain):
+        """Return the record from the domain used in `message_fetch`."""
+        model = res_id = None
+        for part in domain:
+            if len(part) == 3 and part[0] == "model" and part[1] == "=":
+                model = part[2]
+            if len(part) == 3 and part[0] == "res_id" and part[1] == "=":
+                res_id = part[2]
```
Comment (mmequignon) [inline mail_thread_create_nolog/models/mail_message.py:63]: you could probably break or return here, once both model / res_id are set.

File: mail_thread_create_nolog/models/mail_thread.py (hunk 3 above)
Comment (sebalix) [reply mail_thread_create_nolog/models/mail_thread.py:16]: Yes I was thinking the same...

File: mail_thread_create_nolog/models/mail_message.py (hunk 6 above)
Comment (sebalix) [reply mail_thread_create_nolog/models/mail_message.py:53]: Yes, Odoo client doesn't complain

File: mail_thread_create_nolog/__manifest__.py (hunk 5 above)
Comment (sebalix) [reply mail_thread_create_nolog/__manifest__.py:4]: From user/admin POV it is a fake log (which replaces the real log in the DB), and to not duplicate information on UI the real log is of course not created through `mail_create_nolog` context key.
But other feedbacks are welcomed, it's not easy to name a module and give it a good name/description.

File: mail_thread_create_nolog/models/mail_message.py (hunk 6 above)
Comment (sebalix) [reply mail_thread_create_nolog/models/mail_message.py:53]: Improved the comment

File: mail_thread_create_nolog/models/mail_message.py (hunk 7 above)
Comment (sebalix) [reply mail_thread_create_nolog/models/mail_message.py:63]: Done

**Comment by sebalix** [comment]
I added a fix to check if the data model of the record has the `_log_access` enabled (so we are sure it has `create_uid`/`create_date` fields), `queue.job` has this option disabled for instance.

**Comment by simahawk** [comment]
@sebalix good to go?

**Comment by sebalix** [comment]
@sebalix good to go, in production since a while now, no issue so far.

**Pull request merged**